monkey/monkey_island/cc/services/agent_signals_service.py

```diff
@@ -28,18 +28,19 @@ def get_signals(self, agent_id: AgentID) -> AgentSignals:
         return AgentSignals(terminate=terminate_timestamp)
 
     def _get_terminate_signal_timestamp(self, agent_id: AgentID) -> Optional[datetime]:
+        agent = self._agent_repository.get_agent_by_id(agent_id)
         simulation = self._simulation_repository.get_simulation()
         terminate_all_signal_time = simulation.terminate_signal_time
-        if terminate_all_signal_time is None:
-            return None
+        progenitor = self._agent_repository.get_progenitor(agent)
 
-        agent = self._agent_repository.get_agent_by_id(agent_id)
-        if agent.start_time <= terminate_all_signal_time:
-            return terminate_all_signal_time
+        if agent.stop_time is not None:
+            return agent.stop_time
 
-        progenitor = self._agent_repository.get_progenitor(agent)
-        if progenitor.start_time <= terminate_all_signal_time:
-            return terminate_all_signal_time
+        if terminate_all_signal_time is not None:
+            if (agent.start_time <= terminate_all_signal_time) or (
+                progenitor.start_time <= terminate_all_signal_time
+            ):
+                return terminate_all_signal_time
```

Comment:
I think this is harder to read than the old logic. It's nested and there's a compound conditional.

Not that it matters, practically speaking, but it's also less efficient. With the old logic, you rarely need to get the progenitor. With the new logic, you always need to retrieve it.